src/cdk/v2/destinations/intercom/utils.js

```diff
@@ -0,0 +1,254 @@
+  const data = JSON.stringify({
+    query: {
+      operator: 'AND',
+      value: [
+        {
+          field: lookupField,
+          operator: '=',
+          value: lookupFieldValue,
+        },
+      ],
+    },
+  });
+
+  const headers = getHeaders(destination);
+  const baseEndPoint = getBaseEndpoint(destination);
+  const endpoint = `${baseEndPoint}/${SEARCH_CONTACT_ENDPOINT}`;
+  const response = await httpPOST(endpoint, data, {
+    headers,
+    destType: 'intercom',
+    feature: 'transformation',
+  });
+  const processedUserResponse = processAxiosResponse(response);
+  if (isHttpStatusSuccess(processedUserResponse.status)) {
+    return processedUserResponse.response?.data.length > 0
+      ? processedUserResponse.response?.data[0]?.id
+      : null;
+  }
+
+  throw new NetworkError(
+    `Unable to search contact due to : ${JSON.stringify(processedUserResponse?.response?.errors)}`,
+    processedUserResponse?.status,
+    {
+      [tags]: getDynamicErrorType(processedUserResponse?.status),
+    },
+    processedUserResponse,
+  );
+};
+
+/**
+ * Api call to create or update companies in intercom
+ * Ref doc : https://developers.intercom.com/docs/references/rest-api/api.intercom.io/Companies/createOrUpdateCompany/
+ * @param {*} payload
+ * @param {*} destination
+ * @returns
+ */
+const createOrUpdateCompany = async (payload, destination) => {
+  const headers = getHeaders(destination);
+  const finalPayload = JSON.stringify(removeUndefinedAndNullValues(payload));
+  const baseEndPoint = getBaseEndpoint(destination);
+  const endpoint = `${baseEndPoint}/${CREATE_OR_UPDATE_COMPANY_ENDPOINT}`;
+  const response = await httpPOST(endpoint, finalPayload, {
+    headers,
+    destType: 'intercom',
+    feature: 'transformation',
+  });
+
+  const processedResponse = processAxiosResponse(response);
+  if (isHttpStatusSuccess(processedResponse.status)) {
+    return processedResponse.response?.id;
+  }
+
+  throw new NetworkError(
+    `Unable to Create or Update Company due to : ${JSON.stringify(
+      processedResponse?.response?.errors,
+    )}`,
+    processedResponse?.status,
+    {
+      [tags]: getDynamicErrorType(processedResponse?.status),
+    },
+    processedResponse,
+  );
+};
+
+/**
+ * Returns metadata object
+ * @param {*} metadata
+ * @returns
+ */
+const separateReservedAndRestMetadata = (metadata) => {
+  const reservedMetadata = {};
+  const restMetadata = {};
+  if (metadata) {
+    Object.entries(metadata).forEach(([key, value]) => {
+      if (value && typeof value === 'object') {
+        const hasMonetaryAmountKeys = MetadataTypes.monetaryAmount.every((type) => type in value);
+        const hasRichLinkKeys = MetadataTypes.richLink.every((type) => type in value);
+        if (hasMonetaryAmountKeys || hasRichLinkKeys) {
+          reservedMetadata[key] = value;
+        } else {
+          restMetadata[key] = value;
+        }
+      } else {
+        restMetadata[key] = value;
+      }
```

Comment:
```suggestion
      const hasMonetaryAmountKeys = MetadataTypes.monetaryAmount.every((type) => type in value);
      const hasRichLinkKeys = MetadataTypes.richLink.every((type) => type in value);
      if ((value && typeof value === 'object') && (hasMonetaryAmountKeys || hasRichLinkKeys)) {
          reservedMetadata[key] = value;
      } else {
        restMetadata[key] = value;
      }
```

Reply:
@ItsSudip  above condition will not work as we are doing `type in value`. in operator will fail if value is string or number